jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/paths/PathCollection.java

```diff
@@ -0,0 +1,291 @@
+    {
+    }
+
+    public PathCollection(Path... paths)
+    {
+        clear();
+        addAll(List.of(paths));
+    }
+
+    public void setPaths(Path... paths)
+    {
+        clear();
+        addAll(List.of(paths));
+    }
+
+    private Path toSmartPath(Path path)
+    {
+        if (Files.isDirectory(path))
+            return path;
+
+        if (Files.isRegularFile(path) && path.getFileName().toString().toLowerCase(Locale.ENGLISH).endsWith(".jar"))
+        {
+            try
+            {
+                FileSystem fs = FileSystemPool.acquire(path);
+                return fs.getPath("/");
+            }
+            catch (IOException e)
+            {
+                throw new IllegalStateException("ZipFs failure: " + path, e);
+            }
+        }
+
+        throw new IllegalArgumentException("Unsupported Path: " + path);
+    }
+
+    @Override
+    public boolean add(Path element)
+    {
+        return super.add(toSmartPath(element));
+    }
+
+    @Override
+    public void add(int index, Path element)
+    {
+        super.add(index, toSmartPath(element));
+    }
+
+    @Override
+    public boolean addAll(Collection<? extends Path> c)
+    {
+        boolean ret = false;
+        for (Path p : c)
+        {
+            ret |= add(p);
+        }
+        return ret;
+    }
+
+    @Override
+    public boolean addAll(int index, Collection<? extends Path> c)
+    {
+        int idx = index;
+        for (Path p : c)
+        {
+            add(idx++, p);
+        }
+        return true;
+    }
+
+    @Override
+    public void close() throws Exception
+    {
+        clear();
+    }
+
+    @Override
+    public void trimToSize()
+    {
+        throw new UnsupportedOperationException("Not supported by pathCollection");
+    }
+
+    @Override
+    public Path set(int index, Path element)
+    {
+        Path old = super.set(index, toSmartPath(element));
+        try
+        {
+            FileSystemPool.release(old);
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeException("Unable to release old path: " + old);
+        }
+        return old;
+    }
+
+    @Override
+    public void clear()
+    {
+        RuntimeException error = null;
+
+        for (Path path : this)
+        {
+            try
+            {
+                FileSystemPool.release(path);
+            }
+            catch (IOException e)
+            {
+                if (error == null)
+                    error = new RuntimeException("Unable to release some paths");
+                error.addSuppressed(e);
+            }
+        }
+        if (error != null)
+            throw error;
+    }
+
+    /**
+     * Resolve the first path that exists against the collection of paths.
+     *
+     * @param other the other path to resolve against, assumes a String in unix format, that is already URI decoded from URI space.
+     * @see Path#resolve(String)
+     */
+    public Path resolveFirstExisting(String other)
```

Comment:
I think this can just be called `resolve(String)` to make the usage similar to using `'Path`.

Reply:
Since this resolves from a collection of `Path`s to a single `Path` the `resolveFirst` prefix is quite accurate IMHO